title: "R_assignment"

library(tidyverse)
library(data.table)
library(reshape)
library(dplyr)
```

# Data Inspection

```{r data-inspection}
# Load datasets
fang <- read.delim("C:/Users/risha/Downloads/fang_et_al_genotypes.txt")
fangdim <- dim(fang)
fang_info <- file.info("fang_et_al_genotypes.txt")

snp <- read.delim("C:/Users/risha/Downloads/snp_position (1).txt")
snpdim <- dim(snp)
snp_info <- file.info("snp_position.txt")

fang_t <- read.delim("C:/Users/risha/Downloads/transposed_genotypes.txt")
fang_t <- as.data.frame(fang_t)

# Clean and reformat
new_fang <- fang_t[-c(0,1), ]
colnames(new_fang) <- as.character(new_fang[1, ])
new_fang <- new_fang[-c(1), ]
```

#  Data Processing


```{r data-processing}
# Merge SNP position with genotype data
merged <- merge(snp, new_fang, by.x="SNP_ID", by.y="Group", all=TRUE)
final <- merged[-c(2, 5:15)]  # drop unused columns

# Subset maize and teosinte
maize <- final[c(1, 2, 3, 1213:2785)]
teosinte <- final[c(1, 2, 3, 77:1010, 1166:1206)]

# Replace missing values for incremental and decremental
maize_inc <- maize
maize_inc[maize_inc == "?/?"] <- "?"
inc_chr <- split(maize_inc, maize_inc$Chromosome)
sorted_data_inc <- lapply(inc_chr, function(df) df[order(as.numeric(df$Position)),])
lapply(names(sorted_data_inc), function(chr) {
  write.csv(sorted_data_inc[[chr]], file=paste0("inc_chromosome_", chr, ".txt"), row.names=FALSE)
})

maize_dec <- maize
maize_dec[maize_dec == "?/?"] <- "-"
dec_chr <- split(maize_dec, maize_dec$Chromosome)
sorted_data_dec <- lapply(dec_chr, function(df) df[order(as.numeric(df$Position), decreasing = TRUE),])
lapply(names(sorted_data_dec), function(chr) {
  write.csv(sorted_data_dec[[chr]], file=paste0("dec_chromosome_", chr, ".txt"), row.names=FALSE)
})

# SNP Count Summary
maize_snp_count <- aggregate(SNP_ID ~ Chromosome, data = maize, FUN = length)
teosinte_snp_count <- aggregate(SNP_ID ~ Chromosome, data = teosinte, FUN = length)
colnames(maize_snp_count)[2] <- "SNP_Count"
colnames(teosinte_snp_count)[2] <- "SNP_Count"
maize_snp_count$Group <- "Maize"
teosinte_snp_count$Group <- "Teosinte"
snp_counts <- rbind(maize_snp_count, teosinte_snp_count)

# Map chromosome names to numeric
snp_counts$Chromosome_Numeric <- NA
snp_counts$Chromosome_Numeric[snp_counts$Chromosome == "1"] <- 1
snp_counts$Chromosome_Numeric[snp_counts$Chromosome == "2"] <- 2
snp_counts$Chromosome_Numeric[snp_counts$Chromosome == "10"] <- 10
snp_counts$Chromosome_Numeric[snp_counts$Chromosome == "multiple"] <- 11
snp_counts$Chromosome_Numeric[snp_counts$Chromosome == "unknown"] <- 12

# Order factor levels
unique_chromosomes <- unique(snp_counts[, c("Chromosome", "Chromosome_Numeric")])
unique_chromosomes <- unique_chromosomes[order(unique_chromosomes$Chromosome_Numeric), ]
sorted_chromosome_levels <- unique_chromosomes$Chromosome
snp_counts$Chromosome <- factor(snp_counts$Chromosome, levels = sorted_chromosome_levels)
```

# Visualization

```{r plot-snp-bar}
ggplot(snp_counts, aes(x = Chromosome, y = SNP_Count, fill = Group)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Distribution of SNPs Across Chromosomes",
       x = "Chromosome", y = "Number of SNPs") +
  scale_fill_manual(values = c("Maize" = "#E69F00", "Teosinte" = "#56B4E9")) +
  theme_minimal() +
  theme(legend.position = "top")
```

```{r plot-homozygous}
both_long <- filter(fang, Group %in% c("ZMMIL", "ZMMLR", "ZMMMR", "ZMPBA", "ZMPIL", "ZMPJA"))
both <- melt(as.data.table(both_long), measure.vars = colnames(fang)[4:986])
colnames(both)[4:5] <- c("SNP_ID", "Homozygous")
both <- mutate(both, Homozygous = ifelse(Homozygous %in% c("A/A", "C/C", "G/G", "T/T"), TRUE, 
                                ifelse(Homozygous %in% c("A/C", "A/G", "A/T", "C/G", "C/T", "G/T"), FALSE, NA)))
ggplot(data = both) +
  geom_bar(mapping = aes(x = Group, fill = Homozygous), stat = "count") +
  labs(title = "SNPs across groups", x = "Group", y = "Number of SNPs") +
  theme_minimal()
```

```{r genotype-type-distribution}
fang_long <- pivot_longer(fang, cols = -c(Sample_ID, JG_OTU, Group),
                          names_to = "SNP", values_to = "Genotype")
fang_long <- mutate(fang_long,
  Genotype_Type = case_when(
    Genotype == "?" ~ "Missing",
    str_detect(Genotype, "/") & str_sub(Genotype, 1, 1) == str_sub(Genotype, 3, 3) ~ "Homozygous",
    str_detect(Genotype, "/") & str_sub(Genotype, 1, 1) != str_sub(Genotype, 3, 3) ~ "Heterozygous",
    TRUE ~ "Other"))

summary_data <- fang_long %>%
  filter(Genotype_Type != "Missing") %>%
  group_by(Group, Genotype_Type) %>%
  summarise(Count = n()) %>%
  mutate(Proportion = Count / sum(Count))

ggplot(summary_data, aes(x = Group, y = Proportion, fill = Genotype_Type)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Proportion of Homozygous vs. Heterozygous Genotypes by Group",
       x = "Group", y = "Proportion", fill = "Genotype Type") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
